### ISSUE TITLE
attention_layer.py

### ISSUE THREAD

**Issue by karthik78437** [karthik78437/karthik#3]
"""
attention_layer.py

Implements a time-distributed attention mechanism for Keras.
This computes attention weights across time steps of the LSTM outputs.

Reference: simple additive attention over timesteps.
"""

import tensorflow as tf
from tensorflow.keras.layers import Layer
import tensorflow.keras.backend as K

class TemporalAttention(Layer):
    """
    Temporal (time) attention layer.

    Inputs:
        - RNN outputs: shape (batch_size, time_steps, hidden_dim)
    Outputs:
        - context vector: shape (batch_size, hidden_dim)
        - attention_weights: shape (batch_size, time_steps)
    """

    def __init__(self, return_attention=False, **kwargs):
        self.return_attention = return_attention
        super(TemporalAttention, self).__init__(**kwargs)

    def build(self, input_shape):
        # input_shape = (batch_size, time_steps, hidden_dim)
        self.time_steps = input_shape[1]
        self.hidden_dim = input_shape[2]
        # trainable parameters
        self.W = self.add_weight(name='W_att', shape=(self.hidden_dim, self.hidden_dim),
                                 initializer='glorot_uniform', trainable=True)
        self.v = self.add_weight(name='v_att', shape=(self.hidden_dim,),
                                 initializer='glorot_uniform', trainable=True)
        self.b = self.add_weight(name='b_att', shape=(self.time_steps,),
                                 initializer='zeros', trainable=True)
        super(TemporalAttention, self).build(input_shape)

    def call(self, inputs, mask=None):
        # inputs: (batch, time, hidden)
        # score = v^T * tanh(H_t * W + b)
        # compute H * W -> (batch, time, hidden)
        H_W = K.dot(inputs, self.W)  # (b, t, h)
        # tanh(H_W) -> (b, t, h)
        score_t = K.tanh(H_W)  # (b, t, h)
        # dot with v -> (b, t)
        score = K.dot(score_t, K.expand_dims(self.v))  # (b, t, 1)
        score = K.squeeze(score, axis=-1)
        # add bias (broadcasted on time dimension)
        score = score + self.b
        # attention weights
        alpha = K.softmax(score, axis=1)  # (b, t)
        # context vector: sum alpha_t * H_t
        alpha_expanded = K.expand_dims(alpha, axis=-1)  # (b, t, 1)
        context = K.sum(alpha_expanded * inputs, axis=1)  # (b, hidden)
        if self.return_attention:
            return [context, alpha]
        return context

    def compute_output_shape(self, input_shape):
        if self.return_attention:
            return [(input_shape[0], input_shape[2]), (input_shape[0], input_shape[1])]
        return (input_shape[0], input_shape[2])

    def get_config(self):
        cfg = super(TemporalAttention, self).get_config()
        cfg.update({"return_attention": self.return_attention})
        return cfg
